spec/gutter-spec.coffee

```diff
@@ -0,0 +1,74 @@
+{Gutter, DEFAULT_PRIORITY} = require '../src/gutter'
+
+describe 'Gutter initialization tests.', ->
+  name = 'name'
+  fakeGutterContainer = {}
+
+  it 'Test initializing a Gutter with all options.', ->
+    location = 'left'
+    priority = 1
+    visibleInitially = false
+    options =
+      'location': 'left'
+      'priority': 1
+      'visibleInitially': false
+    gutter = new Gutter name, fakeGutterContainer, options
+    expect(gutter.name).toBe name
+    expect(gutter.gutterContainer).toBe fakeGutterContainer
+    expect(gutter.location).toBe location
+    expect(gutter.priority).toBe priority
+    expect(gutter.visible).toBe visibleInitially
+
+  it 'Test initializeing a Gutter with minimal options.', ->
+    fakeGutterContainer = {};
+    gutter = new Gutter name, fakeGutterContainer
+    expect(gutter.name).toBe name
+    expect(gutter.gutterContainer).toBe fakeGutterContainer
+    expect(gutter.location).toBe 'left' # default
+    expect(gutter.priority).toBe DEFAULT_PRIORITY
+    expect(gutter.visible).toBe true # default
+
+describe 'Test hiding and showing a gutter.', ->
+  fakeGutterContainer = {}
+  fakeListener =
+    onGutterShow: null
+    onGutterHide: null
+  name = 'name'
+
+  attachListenerToGutter = (gutter) ->
+    gutter.onShow () ->
+      fakeListener.onGutterShow()
+    gutter.onHide () ->
+      fakeListener.onGutterHide()
+    spyOn fakeListener, 'onGutterShow'
+    spyOn fakeListener, 'onGutterHide'
+
```

Comment:
I want to use the spec below to talk about a few things.

* I think using `::onDidChangeVisible` instead of `::onHide` and `::onShow`
  would yield more consistency with the panel API.
* We're avoiding property access from outside objects in the API except in
  very specific circumstances, such as properties on the `atom` global. So we
  should test against an `::isVisible()` method instead of the `::visible` 
  property.
* Using ad-hoc state like the `events` array below can often be simpler than
  setting up spies. If you still want to use spies to test callbacks, you should
  use `jasmine.createSpy` to create an anonymous function that you pass in,
  as follows:
  ```coffee
  didChangeVisible = jasmine.createSpy('didChangeVisible')
  gutter.onDidChangeVisible(didChangeVisible)
  # do stuf...
  expect(didChangeVisible).toHaveBeenCalled()
  ```
  But like I said, I often prefer a more ad-hoc approach like that used below
  to messing around with spies for callbacks.
* I think the `Gutter` should just take a single params hash since we're already
  supporting options.
* I think the `visibleInitially` option should simply be called `visible` since
  you're already passing it to an initializer.

```coffee
describe "::hide", ->
  it "hides the gutter if it is visible", ->
    events = []
    gutter = new Gutter(name: "test-gutter", container: fakeContainer, visible: true)
    gutter.onDidChangeVisible (visible) -> events.push(visible)
    
    expect(gutter.isVisible()).toBe true
    gutter.hide()
    expect(gutter.isVisible()).toBe false
    expect(events).toEqual [false]
    gutter.hide()
    expect(gutter.isVisible()).toBe false
    expect(events).toEqual [false] # no event if gutter is already hidden
```